Add `hatch`, `ruff`, `pre-commit` and dev convenience (#7409)
yt-dlp/yt-dlp: CONTRIBUTING.md

# DEVELOPER INSTRUCTIONS

Most users do not need to build yt-dlp and can [download the builds](https://github.com/yt-dlp/yt-dlp/releases), get them via [the other installation methods](README.md#installation) or directly run it using `python -m yt_dlp`.

`yt-dlp` uses [`hatch`](<https://hatch.pypa.io>) as a project management tool.
You can easily install it using [`pipx`](<https://pipx.pypa.io>) via `pipx install hatch`, or else via `pip` or your package manager of choice.

If you plan on contributing to `yt-dlp`, you are required to run

```shell
$ hatch run setup
```

> to install a `pre-commit` hook so that required checks (linting, formatting) will run automatically before each commit. If any of the checks fail, then the commit will be blocked; you'll need to fix the failing case and then you can commit the fixed version.

Reviewer: need to update this section now

something like this?
```suggestion
to install a `pre-commit` hook so that required checks/fixes (linting, formatting) will run automatically before each commit. If any code needed to be linted or formatted, then the necessary changes will be made, but the commit will be blocked; you can review the edits and re-commit the fixed version.
```

> so that required checks run automatically before a commit (using pre-commit). This will ensure that certain tests are running before a commit, like formatting. A failing test will mean that the commit will be blocked; fix the failing case and try to commit the fixed version again.

Reviewer: just removing some redundancy / changing some phrasing:
```suggestion
to install a `pre-commit` hook so that required checks (linting, formatting) will run automatically before each commit. If any of the checks fail, then the commit will be blocked; you'll need to fix the failing case and then you can commit the fixed version.
```

After this you can use `hatch shell` to enable a virtual environment that has development dependencies as well as `yt-dlp` installed.

Scripts can be used to run simple tasks, like linting or testing, without having to run `hatch shell` first:

> ```shell
> $ hatch run lint
> $ hatch run format
> $ hatch run test
> ```

Author: Add note that people should run `hatch run format` before `hatch run lint`?
````suggestion
```shell
$ hatch run format
$ hatch run lint
$ hatch run test
```

Make sure to run `hatch run format` before `hatch run lint` since `autopep8` might fix errors that `ruff` would detect but not fix.
````

> After this you can use `hatch shell` to enable a virtual environment that has development dependencies as well as `yt-dlp` installed. Scripts can be used to run simple tasks, like linting or testing, without having to run `hatch shell` first:

Reviewer: phrasing / clarity
```suggestion
After this you can use `hatch shell` to enable a virtual environment that has `yt-dlp` and its development dependencies installed.

In addition, the following script commands can be used to run simple tasks such as linting or testing (without having to run `hatch shell` first):
```

> * `hatch run fix`: Both format and lint the code

Reviewer: ```suggestion
* `hatch run fix`: Automatically fix linter violations and apply required code formatting changes
```

> To automatically fix linter violations as well as format the code, use `hatch run fix`.

Reviewer: this could be consolidated into the above list IMO
```suggestion
```

# DEVELOPER INSTRUCTIONS

Most users do not need to build yt-dlp and can [download the builds](https://github.com/yt-dlp/yt-dlp/releases), get them via [the other installation methods](README.md#installation) or directly run it using `python -m yt_dlp`.

`yt-dlp` uses [`hatch`](<https://hatch.pypa.io>) as a project management tool.
You can easily install it using [`pipx`](<https://pipx.pypa.io>) via `pipx install hatch`, or else via `pip` or your package manager of choice.

If you plan on contributing to `yt-dlp`, you are required to run

```shell
$ hatch run install
```

before creating commits so that the pre-commit hooks are properly initialized.
After this you can use `hatch shell` to enable a virtual environment that has development dependencies as well as `yt-dlp` installed.

Scripts can be used to run simple tasks, like linting or testing, without having to run `hatch shell` first:

```shell
$ hatch run format

> $ hatch run lint

Author: `hatch run fix` runs both now. Im debating if this code block should be included at all since it makes it seem like a list of commands to run; instead it should just show what important `hatch run` commands exist

# DEVELOPER INSTRUCTIONS

Most users do not need to build yt-dlp and can [download the builds](https://github.com/yt-dlp/yt-dlp/releases), get them via [the other installation methods](README.md#installation) or directly run it using `python -m yt_dlp`.

`yt-dlp` uses [`hatch`](<https://hatch.pypa.io>) as a project management tool.
You can easily install it using [`pipx`](<https://pipx.pypa.io>) via `pipx install hatch`, or else via `pip` or your package manager of choice. Make sure you are using at least version `1.10.0`, otherwise some functionality might not work as expected.

If you plan on contributing to `yt-dlp`, best practice is to start by running the following command:

```shell
$ hatch run setup
```

> The above command will install a `pre-commit` hook so that required checks/fixes (linting, formatting) will run automatically before each commit. If any code needs to be linted or formatted, then the necessary changes will be made, but the commit will be blocked; you can review the edits and re-commit the fixed version.

Reviewer 2: I think:

- Those new necessary changes will be added **after** the current commit is blocked.
- Accepting the suggestions from the linter and/or formatter is necessary.
- Suggested changes will be mixed with user's ones, so reviewing all changes is necessary.

So:

```suggestion
The above command will install a `pre-commit` hook so that required checks/fixes (linting, formatting) will run automatically before each commit. If any code needs to be linted or formatted, the commit will be blocked, and necessary changes will be made; you should review all edits and re-commit the fixed version.
```

.

Most users do not need to build yt-dlp and can [download the builds](https://github.com/yt-dlp/yt-dlp/releases) or get them via [the other installation methods](README.md#installation).

To quickly run yt-dlp as a developer, you don't need to build anything either. Simply execute

```shell
$ python -m yt_dlp
```

To properly interact with `yt-dlp` as a developer, you should use [`hatch`](<https://hatch.pypa.io>) (easily installed using [`pipx`](<https://pipx.pypa.io>)).
After installing, you can use `hatch shell` to create a shell with all dependencies as well as `yt-dlp` installed.

If you plan on contributing to `yt-dlp`, run

```shell

> $ hatch run install

Reviewer: the only thing this does is install `pre-commit` (for now), right? IMO this part is a bit confusing

Author: Yes but it should be installed when contributing. We might potentially add more things later (though i doubt it)

Author: Oh and it creates the environment in the first place, installing the dependencies, and making subsequent calls a lot faster

Author: I dont think clarifying it as optional will help the cause. That just means people will not have to spend the time checkig it themselves and we will have to explain it for them.

I think explaining what the pre-commit hook does is the better approach. I tried to give a hint using the ` (run  hatch run fix  if it fails)` addon but not sure if and to what extend we should explain it in the contribution guidelines

Author: I think the new docs changes explain pre-commit in enough detail

            }
    ```
1. Add an import in [`yt_dlp/extractor/_extractors.py`](yt_dlp/extractor/_extractors.py). Note that the class name must end with `IE`.

> 1. Run `hatch run test YourExtractor`. This *may fail* at first, but you can continually re-run it until you're done. Upon failure, it will output the missing fields and/or correct values which you can copy. If you decide to add more than one test, the tests will then be named `YourExtractor`, `YourExtractor_1`, `YourExtractor_2`, etc. Note that tests with an `only_matching` key in the test's dict are not included in the count. You can also run all the tests in one go with `YourExtractor_all`

Reviewer: ```suggestion
1. Run `hatch test YourExtractor`. This *may fail* at first, but you can continually re-run it until you're done. Upon failure, it will output the missing fields and/or correct values which you can copy. If you decide to add more than one test, the tests will then be named `YourExtractor`, `YourExtractor_1`, `YourExtractor_2`, etc. Note that tests with an `only_matching` key in the test's dict are not included in the count. You can also run all the tests in one go with `YourExtractor_all`
```

right?
